### PR TITLE
Fix issue of comparing instance URLs with different project ID representations in GCE target pools

### LINKED ISSUE #11327: GCE UpdateTCPLoadBalancer removes all VMs from load balancer
UpdateTCPLoadBalancer works by getting the list of the current nodes in the load balancer's target pool and then comparing them to the list of nodes in the cluster. In the process, though, it compares the project ID returned by the target pool API with the string ID known to the controller manager, but one of them is the numeric project ID and the other is the string project ID. We then add the ones that we know about before removing the ones that look wrong to us. This causes all cluster resizes and upgrades to remove all nodes from existing target pools.

We should avoid comparing the project IDs, since we already know the project ID should be correct given that we used it to get the target pool in the first place.

### PR DESCRIPTION
**Pull request by a-robinson**
Fixes #11327 

I'm working on adding a unit test

@brendandburns @roberthbailey

### CODE REVIEW

**Comment by a-robinson** [comment]
Updated the commit with poor man's unit tests for comparing host links rather than for the UpdateTCPLoadBalancer function as a whole.

File: pkg/cloudprovider/gce/gce.go [hunk 1]
```diff
@@ -225,6 +225,16 @@ func makeHostLink(projectID, zone, host string) string {
 		projectID, zone, host)
 }
 
+func makeComparableHostLink(zone, host string) string {
+	host = canonicalizeInstanceName(host)
+	return fmt.Sprintf("/zones/%s/instances/%s", zone, host)
+}
+
+func makeHostLinkComparable(hostLink string) string {
+	idx := strings.Index(hostLink, "/zones/")
```
Comment (roberthbailey) [inline pkg/cloudprovider/gce/gce.go]: check for idx < 0? Otherwise you'll cause a panic.

**Comment by brendandburns** [comment]
LGTM.  @thockin @roberthbailey can one of you take a quick look too?

**Comment by erictune** [comment]
I'm not merging any code until Jenkins is green for a turn.

**Comment by a-robinson** [comment]
Jenkins is green right now, but I'm going to add the < 0 check

File: pkg/cloudprovider/gce/gce.go (hunk 1 above)
Comment (a-robinson) [inline pkg/cloudprovider/gce/gce.go]: Fixed

**Comment by a-robinson** [comment]
Sorry, when I said it was green, I meant the run specific to this PR was green. I hadn't checked kubekins.

**Comment by roberthbailey** [comment]
LGTM.

File: pkg/cloudprovider/gce/gce.go [hunk 2]
```diff
@@ -225,6 +225,19 @@ func makeHostLink(projectID, zone, host string) string {
 		projectID, zone, host)
 }
 
+func makeComparableHostLink(zone, host string) string {
```
Comment (thockin) [inline pkg/cloudprovider/gce/gce.go]: These two functions have teh same name, but the words are in different order.  Can we do better?

**Comment by thockin** [comment]
naming nit then LGTM

**Comment by a-robinson** [comment]
Renamed things

File: pkg/cloudprovider/gce/gce.go [hunk 3]
```diff
@@ -219,12 +219,25 @@ func (gce *GCECloud) Routes() (cloudprovider.Routes, bool) {
...
 	return fmt.Sprintf("https://www.googleapis.com/compute/v1/projects/%s/zones/%s/instances/%s",
 		projectID, zone, host)
 }
 
+func makeComparableHostPath(zone, host string) string {
+	host = canonicalizeInstanceName(host)
+	return fmt.Sprintf("/zones/%s/instances/%s", zone, host)
+}
+
+func hostURLToComparablePath(hostURL string) string {
```
Comment (thockin) [inline pkg/cloudprovider/gce/gce.go]: for future, this is a great place to use typedefs (I don't know what Go actuallys calls them) to annotate the contents and semantics of a string value.

**Comment by thockin** [comment]
LGTM  I assume this warrants a cherry pick

**Comment by roberthbailey** [comment]
Yes. Into the 0.21-release branch as well as the 1.0-release branch.

**Pull request merged**